### ISSUE TITLE
BUG: test failure on ppc64le

### ISSUE THREAD

**Issue by zhsj** [seccomp/libseccomp-golang#61]
```
$ go test -v . -run TestRuleAddAndLoad
=== RUN   TestRuleAddAndLoad
    seccomp_test.go:588: Syscall should have returned error code!
--- FAIL: TestRuleAddAndLoad (0.00s)
```

https://github.com/seccomp/libseccomp-golang/blob/847368b35ebf24980814276c545776d22f66d71e/seccomp_test.go#L586-L588

The pid on ppc64le is `1`, don't know why it's not minus..

libseccomp is 2.5.0, and libseccomp-golang is head version.

**Comment by pcmoore** [comment]
Hi @zhsj, thanks for the report, I'm sorry for the delay in responding.

Unfortunately I don't have a ppc64le system to test this, would you be able to debug this further?  @drakenclimber and I may be able to help guide the debugging if needed ...

**Comment by zhsj** [comment]
> Hi @zhsj, thanks for the report, I'm sorry for the delay in responding.
> 
> Unfortunately I don't have a ppc64le system to test this, would you be able to debug this further? @drakenclimber and I may be able to help guide the debugging if needed ...

Thanks, I do need some guides since I don't know much about it.

From my observation, seccomp is effected in this test, since syscall.Getpid returns `1` instead of the real pid. But I can't tell why it's `1` instead of `-1`.  

In https://github.com/seccomp/libseccomp-golang/blob/847368b35ebf24980814276c545776d22f66d71e/seccomp_test.go#L553

It says to return `0x1`. If changing it to `0x2`, then on ppc64le, the syscall.Getpid returns `2`, but on amd64, it's `-2`. Do you know what's reason that syscall has negate value  here?

**Comment by pcmoore** [comment]
Hmm, the difference between ppc64le and x86-64 is interesting.  Any chance you could try a similar filter using the native/C libseccomp API?  I'm curious if this is an oddity in the kernel/libc or if the filter is somehow to blame.

The C based filter should be trivial, but if you are not comfortable writing one let me know and we can provide a simple test program to compile and run.

**Comment by zhsj** [comment]
```
(sid_ppc64el-dchroot)zhsj@plummer:~/t$ cat test.c 
#include <seccomp.h>
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

int main() {
  pid_t pid;
  pid = getpid();
  printf("pid: %d\n", pid);
  scmp_filter_ctx ctx;
  ctx = seccomp_init(SCMP_ACT_ALLOW);
  seccomp_rule_add(ctx, SCMP_ACT_ERRNO(0x1), SCMP_SYS(getpid), 0);
  seccomp_load(ctx);
  pid = getpid();
  printf("pid: %d\n", pid);
  return 0;
}

(sid_ppc64el-dchroot)zhsj@plummer:~/t$ gcc test.c -lseccomp
(sid_ppc64el-dchroot)zhsj@plummer:~/t$ ./a.out 
pid: 27256
pid: 1
(sid_ppc64el-dchroot)zhsj@plummer:~/t$ dpkg -l|grep seccomp
ii  libseccomp-dev:ppc64el           2.5.0-3                        ppc64el      high level interface to Linux seccomp filter (development files)
ii  libseccomp2:ppc64el              2.5.0-3                        ppc64el      high level interface to Linux seccomp filter
(sid_ppc64el-dchroot)zhsj@plummer:~/t$ 
```

**Comment by zhsj** [comment]
Tried with libseccomp master, same result.

**Comment by pcmoore** [comment]
Hi @zhsj.  First let me apologize for the delay in responding - things have been busy.

I tried your reproducer on my x86_64 system and run into the same result as you (on your x86_64 tests):
```
% gcc -o test test.c -lseccomp
% ./test 
pid: 10140
pid: -1
% pacman -Q libseccomp
libseccomp 2.5.1-2
```

I'll have to take a look and get back to you on this, it definitely looks to be like a problem in the main libseccomp code base.

@drakenclimber have you seen anything like this?

**Comment by pcmoore** [comment]
FWIW, I see similar results on x86_64 with the `openat()` syscall.

I'm beginning to wonder if this is simply a difference in the seccomp handling for each architecture; x86_64 returns the errno value negated whereas ppc64le doesn't negate the errno value.  More investigation is needed.

**Comment by drakenclimber** [comment]
> @drakenclimber have you seen anything like this?

I haven't, but I will look into it this week.

**Comment by drakenclimber** [comment]
I happened to have access to an s390x machine, so I ran the above test program on it.  Looks like s390x behaves like x86_64.  It's feeling like a ppc64le-specific issue to me.

```
$ uname -r
4.18.0-240.el8.s390x

$ ./a.out 
pid: 8599
pid: -1

```

**Comment by drakenclimber** [comment]
Ewww....

I dug up my old account to some IBM ppc cloud boxes.  On RHEL8, I saw the same behavior @zhsj reported in https://github.com/seccomp/libseccomp-golang/issues/61#issuecomment-737346374 above.  But I had troubles with IBM's yum repos - I couldn't get gperf installed - so I destroyed the VM.

Here's what I saw on an RHEL7 ppc64le box.

```
$ uname -r
3.10.0-1160.el7.ppc64le
$ cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.9 (Maipo)
$ rpm -qai libseccomp
Name        : libseccomp
Version     : 2.3.1
...

$ ./a.out 
pid: 20930
pid: 20930
```

**Comment by drakenclimber** [comment]
And for completeness, it behaved the same with the head of master in libseccomp

```
$ ./a.out 
pid: 26242
Running HEAD of libseccomp
pid: 26242
```

**Comment by drakenclimber** [comment]
I spent some more time on this today.  It looks like libseccomp and seccomp in the kernel are doing the right thing.

On my RHEL8.3 ppc64le box, I again ran the example C program in this [comment](https://github.com/seccomp/libseccomp-golang/issues/61#issuecomment-737346374).

Kernel info
```
**$ uname -r
4.18.0-240.1.1.el8_3.ppc64le
```

Then I attached a bpftrace script to the output of [seccomp_run_filters()](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/seccomp.c#n391).  This the kernel function that actually runs the BPF.  Here's the output from the test program
```
$ sudo bpftrace -e 'kretprobe:seccomp_run_filters { printf("seccomp_run_filters() returned: 0x%x\n", retval); }'
modprobe: FATAL: Module kheaders not found in directory /lib/modules/4.18.0-240.1.1.el8_3.ppc64le
Attaching 1 probe...
seccomp_run_filters() returned: 0x50001
```

The data portion (0x50001 & 0xFFFF) is 0x1 as was requested.  And the action portion (0x50001 & 0xFFFF0000) was 0x50000 which maps to SECCOMP_RET_ERRNO, again as requested.

In ppc64le, seccomp negates this error code and stores it in gpr[3].  Accessing the RedHat kernel source is somewhat finicky, so I looked through the latest upstream kernel.  It looks like gpr[3] should be properly propagated all the way back to the user.

Unfortunately I didn't have a chance to try out the filter on the latest upstream kernel.

tl;dr - I think libseccomp and seccomp in the kernel are doing the right thing.  This again seems like a RHEL on ppc64le issue.

**Comment by pcmoore** [comment]
It almost looks like the libc is changing the sign on the errno on ppc64le, or am I missing something?

If that is the case (libc munging the errno), I'm not sure we want to change libseccomp as different libc implementations could do different things and I'm not sure how we would detect that at runtime.  Perhaps this is just a note in the docs?

**Comment by drakenclimber** [comment]
> It almost looks like the libc is changing the sign on the errno on ppc64le, or am I missing something?

I think that's a reasonable guess, but I didn't have time to track it down to that.  The kernel code flow in ppc64le is straightforward.

`seccomp_run_filters()` ran the BPF and generated 0x50001. 
`syscall_set_return_value()` is an arch-specific function, and in this example on ppc64le, it correctly set gpr[3] to -1.
After that, nothing in the ppc64le kernel syscall stack modified gpr[3] as best as I could tell.

> 
> If that is the case (libc munging the errno), I'm not sure we want to change libseccomp as different libc implementations could do different things and I'm not sure how we would detect that at runtime. Perhaps this is just a note in the docs?

Could be.  It seems like a strange design choice, but who knows :)

**Comment by drakenclimber** [comment]
Good call on ppc64le's libc.  I updated the test program as follows.  (Note the `syscall()` call instead of `getpid()`).

```
#include <errno.h>
#include <seccomp.h>
#include <stdio.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <unistd.h>

int main() {
  pid_t pid;
  //pid = getpid();
  pid = syscall(SYS_getpid);
  printf("pid: %d\n", pid);
  scmp_filter_ctx ctx;
  ctx = seccomp_init(SCMP_ACT_ALLOW);
  seccomp_rule_add(ctx, SCMP_ACT_ERRNO(0x7), SCMP_SYS(getpid), 0);
  seccomp_load(ctx);
  //pid = getpid();
  pid = syscall(SYS_getpid);
  if (pid < 0)
          printf("errno = %d\n", errno);
  printf("pid: %d\n", pid);
  return 0;
}
```
And the results once it's run.
```
$ ./a.out 
pid: 44982
errno = 7
pid: -1
```

I think a documentation note is the right answer.

**Comment by drakenclimber** [comment]
I opened a libseccomp documentation issue to track this.

**Comment by kolyshkin** [comment]
The test case still fails on PPC, thus reopening.

**Comment by pcmoore** [comment]
As a quick FYI, this was the doc "fix" we applied in the main libseccomp project: https://github.com/seccomp/libseccomp/pull/333

**Comment by pcmoore** [comment]
It may be that the best solution it to simply choose another syscall for the test.

**Comment by kolyshkin** [comment]
> It may be that the best solution it to simply choose another syscall for the test.

Yes, that's what I've done in #100, PTAL

**Comment by pcmoore** [comment]
I just merged PR #100 so I'm going to mark this as closed, but if anyone still notices a problem feel free to comment/reopen.
